### PR TITLE
petsc: simplify dependencies with hypre

### PR DESCRIPTION
**Pull request by tristan0x**
* split in independent `depends_on` directives
* add propagation of `complex` variant to hypre

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -246,24 +246,17 @@ def check_fortran_compiler(self):
...
-    )
-    depends_on(
-        "hypre@2.14:2.22.0+mpi~internal-superlu+int64", when="@3.14:3.15+hypre+mpi~complex+int64"
-    )
-    depends_on("hypre@2.14:+mpi~internal-superlu~int64", when="@3.16:+hypre+mpi~complex~int64")
-    depends_on("hypre@2.14:+mpi~internal-superlu+int64", when="@3.16:+hypre+mpi~complex+int64")
-    depends_on("hypre@develop+mpi~internal-superlu+int64", when="@main+hypre+mpi~complex+int64")
-    depends_on("hypre@develop+mpi~internal-superlu~int64", when="@main+hypre+mpi~complex~int64")
+    depends_on("hypre@:2.13~internal-superlu", when="@:3.8+hypre")
+    depends_on("hypre@2.14:2.18.2~internal-superlu", when="@3.9:3.13+hypre")
```
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: Hm  - spack now lists petsc@3.11: - so perhaps some of these old version dependencies can be removed. I guess thats a different PR [for with fixes for all pkgs - not just hypre]

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -246,24 +246,17 @@ def check_fortran_compiler(self):
...
+    depends_on("hypre@2.14:2.22.0~internal-superlu", when="@3.14:3.15+hypre")
+    depends_on("hypre@2.14:~internal-superlu", when="@3.16:+hypre")
+    depends_on("hypre@develop~internal-superlu", when="@main+hypre")
+    depends_on("hypre+complex", when="+hypre+complex")
+    depends_on("hypre~complex", when="+hypre~complex")
+    depends_on("hypre+int64", when="+hypre+int64")
+    depends_on("hypre~int64", when="-hypre~int64")
```
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: Typo?

```suggestion
    depends_on("hypre~int64", when="+hypre~int64")
```

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 1 above)
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: I think we should modify one of them and retain as:
```
    depends_on("hypre@2.14:2.18.2~internal-superlu", when="@3.11:3.13+hypre")
```

Comment (tristan0x) [reply var/spack/repos/builtin/packages/petsc/package.py]: you are right, I should not have removed the 2 lines. Should be good now with 8221af9
Thank you

**Review by balay — APPROVED** [review]

**Review by bernhardkaindl — APPROVED** [review]
Giving approval after review with 3 changes and approval by @balay as well as successful build on Ubuntu 20.04

**Pull request merged**